skbio/tree/_nni.py

```diff
@@ -110,19 +123,28 @@ def nni(tree, dm, allow_edge_estimation=True, inplace=True):
     for node in tree.non_tips():
         if len(node.children) != 2:
             raise TypeError("Could not perform NNI. Tree needs to be a binary tree.")
-    adm = _average_distance_matrix(tree, dm)
+    if balanced:
+        adm = _balanced_average_matrix(tree, dm)
+    else:
+        adm = _average_distance_matrix(tree, dm)
     while True:
         # create heap of possible swaps and then swapping subtrees
         # until no more swaps are possible.
         adm = _average_distance_matrix(tree, dm)
-        heap = _swap_heap(tree, adm)
+        if balanced:
+            heap = _swap_heap(tree, adm)
+        else:
+            heap = _swap_heap(tree, adm, balanced=False)
         if not heap:
```

Comment:
By changing the order of code, this loop can be modified into `while heap:` in the beginning, without explicitly checking `if not heap:`. But this is trivial and no need to spend time on changing the current code.